src/TestSuite/ContainerStubTrait.php

```diff
@@ -0,0 +1,168 @@
+     */
+    public function configApplication(string $class, ?array $constructorArgs): void
+    {
+        $this->_appClass = $class;
+        $this->_appArgs = $constructorArgs;
+    }
+
+    /**
+     * Create an application instance.
+     *
+     * Uses the configuration set in `configApplication()`.
+     *
+     * @return \Cake\Core\HttpApplicationInterface|\Cake\Core\ConsoleApplicationInterface
+     */
+    protected function createApp()
+    {
+        if ($this->_appClass) {
+            $appClass = $this->_appClass;
+        } else {
+            /** @psalm-var class-string<\Cake\Http\BaseApplication> */
+            $appClass = Configure::read('App.namespace') . '\Application';
+        }
+        if (!class_exists($appClass)) {
+            throw new LogicException("Cannot load `{$appClass}` for use in integration testing.");
+        }
+        $appArgs = $this->_appArgs ?: [CONFIG];
+
+        $app = new $appClass(...$appArgs);
+        if (!empty($this->containerServices) && method_exists($app, 'getEventManager')) {
+            $app->getEventManager()->on('Application.buildContainer', [$this, 'modifyContainer']);
+        }
+        return $app;
+    }
+
+    /**
+     * Add a mocked service to the container.
+     *
+     * When the container is created the provided classname
+     * will be mapped to the factory function. The factory
+     * function will be used to create mocked services.
+     *
+     * @param string $class The class or interface you want to define.
+     * @param \Closure $factory The factory function for mocked services.
+     * @return $this
+     */
+    public function mockService(string $class, Closure $factory)
+    {
+        $this->containerServices[$class] = $factory;
+
+        return $this;
+    }
+
+    /**
+     * Remove a mocked service to the container.
+     *
+     * @param string $class The class or interface you want to remove.
+     * @return $this
+     */
+    public function removeMockService(string $class)
+    {
+        unset($this->containerServices[$class]);
+
+        return $this;
+    }
+
+    /**
+     * Wrap the application's container with one containing mocks.
+     *
+     * If any mocked services are defined, the application's container
+     * will be replaced with one containing mocks. The original
+     * container will be set as a delegate to the mock container.
+     *
+     * @param \Cake\Event\EventInterface $event The event
+     * @param \Cake\Core\ContainerInterface $container The container to wrap.
+     * @return null|\Cake\Core\ContainerInterface
+     */
+    public function modifyContainer(EventInterface $event, ContainerInterface $container): ?ContainerInterface
+    {
+        if (empty($this->containerServices)) {
+            return null;
+        }
+        foreach ($this->containerServices as $key => $factory) {
+            if ($container->has($key)) {
+                $container->extend($key)->setConcrete($factory);
```

Comment:
Does anything implement `extend`? I don't see it anywhere.

Reply:
Yeah it is in League/Container

Reply:
oh, ok, so you're just exposing a League/Container function to the Cake interface that extends the PSR interface.